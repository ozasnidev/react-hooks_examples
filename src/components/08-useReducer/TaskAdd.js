import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TaskAdd = ({ handleAddTask }) => {
    const [{ description }, handleInputChange, resetForm] = useForm({
        description: ''
    });
    const handleSubmit = (event) => {
        event.preventDefault();

        if(description.trim().length >= 1){
            handleAddTask({type: 'add', payload: {id: new Date().getTime(), desc: description, done: false}});
            resetForm();
        }
    };

    return (
        <>
            <div className="col-5">
                <h3>Agregar Task</h3>
                <hr/>
                <form
                    onSubmit={handleSubmit}
                >
                    <input
                        type="text"
                        name="description"
                        className="form-control"
                        placeholder="Task"
                        autoComplete="off"
                        value={ description }
                        onChange={ handleInputChange }
                    />
                    <button
                        type="submit"
                        className="btn btn-outline-success mt-2 btn-block"
                    >
                        Agregar
                    </button>
                </form>
            </div>
        </>
    )
}
