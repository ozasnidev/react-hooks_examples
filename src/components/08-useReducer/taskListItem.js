import React from 'react'

export const TaskListItem = ({task, index, handleDeleteTask, handleToggle}) => {
    return (
        <li
            key={task.id}
            className="list-group-item list-group-item-warning m-1"
        >
            <p  
                className={`${task.done && 'complete'}`}
                onClick={() => handleToggle(task.id)}
            >
                {`${index + 1} ${task.desc}`}
            </p>
            <button 
                className="btn btn-outline-danger"
                onClick={() => handleDeleteTask(task.id)}
            >
                Borrar
            </button>
        </li>
    )
}
