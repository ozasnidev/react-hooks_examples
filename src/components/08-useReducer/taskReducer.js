export const taskReducer = ( state = [], action ) => {
    switch(action.type){


        default: return state;
    }
} 