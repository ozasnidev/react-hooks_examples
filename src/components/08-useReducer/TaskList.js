import React from 'react'
import { TaskListItem } from './taskListItem'

export const TaskList = ({taskState, handleDeleteTask, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
            {
                taskState.map((task, index  ) => (
                    <TaskListItem 
                        key = {task.id}
                        task = {task}   
                        index = {index}
                        handleDeleteTask = {handleDeleteTask}
                        handleToggle = {handleToggle}
                    />
                ))
            }
        </ul>
    )
}
