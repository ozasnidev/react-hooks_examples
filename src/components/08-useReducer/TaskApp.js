import React, { useEffect, useReducer } from 'react'
import { taskReducer } from './taskReducer';
import { TaskList } from './TaskList';
import { TaskAdd } from './TaskAdd';
import './styles.css';

const init = () => {
    return JSON.parse(localStorage.getItem('tasks')) || [];
}

export const TaskApp = () => {
    const [taskState, dispatch] = useReducer(taskReducer, [], init);
    const handleDeleteTask = (taskId) => { dispatch({ type: 'delete', payload: taskId}); };
    const handleToggle = (taskId) => { dispatch({type: 'toggle', payload: taskId}); }
    const handleAddTask = (newTask) => { dispatch(newTask); }
    useEffect( () => {
        localStorage.setItem('tasks', JSON.stringify(taskState));
    }, [taskState]);

    return (
        <>
            <h1>TaskApp ({taskState.length})</h1>
            <hr/>
            <div className="row">
                <div className="col-7">
                    <h3>Task</h3>
                    <hr/>
                    <TaskList 
                        taskState = {taskState}
                        handleDeleteTask = {handleDeleteTask}
                        handleToggle = {handleToggle}
                    />
                </div>
                <TaskAdd
                    handleAddTask = {handleAddTask}
                />
            </div>
        </>
    )
}
