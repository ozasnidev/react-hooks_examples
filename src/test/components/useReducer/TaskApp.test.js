import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { TaskApp } from '../../../components/08-useReducer/TaskApp';
import { mockTask } from '../../fixtures/mockTask';
import { act } from '@testing-library/react';

describe('TaskApp component', () => {
    const wrapper = mount(<TaskApp />);
    Storage.prototype.setItem = jest.fn(() => {});

    test('render', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('add a Task', () => {  
        act(() => { 
            wrapper.find('TaskAdd').prop('handleAddTask')({type: 'add', payload:mockTask[0]});
            wrapper.find('TaskAdd').prop('handleAddTask')({type: 'add', payload:mockTask[1]});
        });
        expect(wrapper.find('h1').text().trim()).toBe('TaskApp (2)');
        expect(localStorage.setItem).toHaveBeenCalledTimes(1);
    });
});

