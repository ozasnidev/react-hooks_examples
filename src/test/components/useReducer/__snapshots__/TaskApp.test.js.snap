// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskApp component render 1`] = `
Array [
  <h1>
    TaskApp (
    0
    )
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <h3>
        Task
      </h3>
      <hr />
      <ul
        className="list-group list-group-flush"
      />
    </div>
    <div
      className="col-5"
    >
      <h3>
        Agregar Task
      </h3>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="description"
          onChange={[Function]}
          placeholder="Task"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-success mt-2 btn-block"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>,
]
`;
